src/rgw/rgw_rest_conn.cc

```diff
@@ -422,12 +501,19 @@ RGWRESTReadResource::RGWRESTReadResource(RGWRESTConn *_conn,
 					 param_vec_t& _params,
 					 param_vec_t *extra_headers,
                                          RGWHTTPManager *_mgr)
-  : cct(_conn->get_ctx()), conn(_conn), resource(_resource), params(_params),
-    cb(bl), mgr(_mgr), req(cct, conn->get_url(), &cb, NULL, NULL, _conn->get_api_name())
+  : cct(_conn->get_ctx()), conn(_conn), url(conn->get_url()), resource(_resource), params(_params),
+    cb(bl), req_ret(0), mgr(_mgr), req(cct, url, &cb, NULL, NULL, _conn->get_api_name())
 {
   init_common(extra_headers);
 }
 
+RGWRESTReadResource::~RGWRESTReadResource()
+{
+  if (req_ret == -EIO) {
+    conn->set_url_unconnectable(url);
+  }
```

Comment:
i'm guessing you're using destructors this way just to avoid duplicating the checks for `send_request()` and `complete_request()`? i'm hoping that we don't need the checks for `send_request()`, and i think it would be clearer if was in the error path itself